src/Store/TitleStoreConnector.re

```diff
@@ -9,65 +9,78 @@ module Model = Oni_Model;
+      switch (Model.Selectors.getActiveBuffer(state)) {
+      | None => initialValues
+      | Some(buf) =>
+        let fp = Model.Buffer.getFilePath(buf);
+        let ret =
+          switch (fp) {
+          | None => initialValues
+          | Some(fp) =>
+            let activeEditorShort = Filename.basename(fp);
+            [("activeEditorShort", activeEditorShort), ...initialValues];
           };
+        switch (Model.Buffer.isModified(buf)) {
+        | false => ret
+        | true => [("dirty", "*"), ...ret]
         };
+      };
 
-      let initialValues =
-        switch (state.workspace) {
-        | None => initialValues
-        | Some(workspace) => [
-            ("rootName", workspace.rootName),
-            ...initialValues,
-          ]
-        };
+    let initialValues =
+      switch (state.workspace) {
+      | None => initialValues
+      | Some(workspace) => [
+          ("rootName", workspace.rootName),
+          ...initialValues,
+        ]
+      };
 
-      initialValues |> List.to_seq |> Core.StringMap.of_seq;
-    };
+    initialValues |> List.to_seq |> Core.StringMap.of_seq;
+  };
 
-  let updateTitleEffect = state =>
-    Isolinear.Effect.createWithDispatch(~name="title.update", _dispatch => {
+module Effects = {
+  let updateTitle = state =>
+    Isolinear.Effect.createWithDispatch(~name="title.update", dispatch => {
       let templateVariables = getTemplateVariables(state);
       let titleTemplate =
         Core.Configuration.getValue(c => c.windowTitle, state.configuration);
 
       let titleModel = Model.Title.ofString(titleTemplate);
       let title = Model.Title.toString(titleModel, templateVariables);
 
+      dispatch(Actions.SetTitle(title));
+    });
+};
+
+let start = setTitle => {
+  let _lastTitle = ref("");
+
+  let internalSetTitleEffect = title =>
+    Isolinear.Effect.createWithDispatch(~name="title.set", _dispatch =>
       if (!String.equal(_lastTitle^, title)) {
         _lastTitle := title;
         setTitle(title);
-      };
-    });
+      }
+    );
 
   let updater = (state: Model.State.t, action: Actions.t) => {
     switch (action) {
-    | Init => (state, updateTitleEffect(state))
-    | BufferEnter(_) => (state, updateTitleEffect(state))
-    | BufferSetModified(_) => (state, updateTitleEffect(state))
-    // Catch directory changes
-    | OpenExplorer(_) => (state, updateTitleEffect(state))
+    | Init => (state, Effects.updateTitle(state))
+    | BufferEnter(_) => (state, Effects.updateTitle(state))
+    | BufferSetModified(_) => (state, Effects.updateTitle(state))
+
+    // TODO: This shouldn't exist, but needs to  be here because it deoends on
+    // `setTitle` being passed in. It would however be ebtter to have a more
+    // general mechanism, like "Effect actions" handled by an injected dependency
+    // or have effects parameterized by an "environment" passed in along with
+    // `dispatch`
+    | SetTitle(title) => (state, internalSetTitleEffect(title))
```

Comment:
:eyes: 

Maybe we should make our own algebraic effect system!

Reply:
@jordwalke had some ideas around user-space algebraic effects!

Reply:
Would be nice to have a better solution for this for sure. I've seen a few articles on Free monads which look interesting: http://degoes.net/articles/modern-fp

It would describe the semantics of effects - like `SetTitle(title)` - but leave it up to an 'interpreter' to decide _what_ to do.

Reply:
I think tagless final might work better with Reason's module system, but yeah something along those lines might be nice. Or overcomplicated. Either way it would be fun to try!